crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1520,6 +1504,16 @@ impl<'db> Type<'db> {
                 .into_callable_type(db)
                 .is_assignable_to(db, target),
 
+            (Type::ProtocolInstance(left), Type::ProtocolInstance(right)) => {
+                left.is_assignable_to(db, right)
+            }
+            // Other than the dynamic types such as `Any`/`Unknown`/`Todo` handled above,
+            // a protocol instance can never be assignable to a nominal type,
+            // with the *sole* exception of `object`.
+            // TODO: `Callable` types are also structural types.
+            (Type::ProtocolInstance(_), _) => false,
+            (_, Type::ProtocolInstance(protocol)) => self.satisfies_protocol(db, protocol),
```

Comment:
There will need to be an argument or a different method to distinguish fully-static `satisfies_protocol` from gradual `satisfies_protocol`.

Reply:
Yes! For now I'll just add a TODO to the doc-comment for `satisfies_protocol()`. There'll need to be lots of changes (which I plan to make incrementally) when we start considering the types of the protocol members.